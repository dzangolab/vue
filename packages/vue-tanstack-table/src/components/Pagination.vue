<template>
  <div class="pagination">
    <div v-if="showItemsPerPageControl" class="items-per-page-control">
      <span>{{ itemsPerPageControlLabel }}</span>
      <select :value="itemsPerPage" @change="handleItemsPerPageChange">
        <option
          v-for="option in itemsPerPageOptions"
          :key="option"
          :value="option"
        >
          {{ option }}
        </option>
      </select>
    </div>

    <div v-if="lastPage > 1 && showPageInput" class="page-input-control">
      <span>{{ pageInputLabel }}</span>
      <DebouncedInput
        type="number"
        :debounce-time="inputDebounceTime"
        @update:model-value="handlePageInputChange"
      />
    </div>

    <div v-if="lastPage > 1" class="pagination-buttons-wrapper">
      <div v-if="showFirstLastButtons || showPreviousNextButtons">
        <ButtonElement
          v-if="showFirstLastButtons"
          :disabled="!(currentPage > 0)"
          class="first-page"
          size="small"
          @click="$emit('update:currentPage', 0)"
        >
          <template #iconLeft>
            <Icon icon="prime:angle-double-left" />
          </template>
        </ButtonElement>
        <ButtonElement
          v-if="showPreviousNextButtons"
          :disabled="!(currentPage > 0)"
          class="previous-page"
          size="small"
          @click="$emit('update:currentPage', currentPage - 1)"
        >
          <template #iconLeft>
            <Icon icon="prime:angle-left" />
          </template>
        </ButtonElement>
      </div>

      <div>
        <template v-if="showPageButtons">
          <ButtonElement
            v-for="page in pages"
            :key="page"
            :class="`page-button ${page === currentPage + 1 ? 'active' : ''}`"
            :label="`${page}`"
            size="small"
            @click="$emit('update:currentPage', page - 1)"
          />
        </template>
        <span v-else>{{ `${currentPage + 1} / ${lastPage}` }}</span>
      </div>

      <div v-if="showFirstLastButtons || showPreviousNextButtons">
        <ButtonElement
          v-if="showPreviousNextButtons"
          :disabled="!(currentPage < lastPage - 1)"
          class="next-page"
          size="small"
          @click="$emit('update:currentPage', currentPage + 1)"
        >
          <template #iconLeft>
            <Icon icon="prime:angle-right" />
          </template>
        </ButtonElement>
        <ButtonElement
          v-if="showFirstLastButtons"
          :disabled="!(currentPage < lastPage - 1)"
          class="last-page"
          size="small"
          @click="$emit('update:currentPage', lastPage - 1)"
        >
          <template #iconLeft>
            <Icon icon="prime:angle-double-right" />
          </template>
        </ButtonElement>
      </div>
    </div>

    <span v-else>{{ `${currentPage + 1} / ${lastPage}` }}</span>
  </div>
</template>

<script lang="ts">
export default {
  name: "TablePagination",
};
</script>

<script setup lang="ts">
import { ButtonElement, DebouncedInput } from "@dzangolab/vue3-ui";
import { Icon } from "@iconify/vue";
import { ref, computed } from "vue";

const props = defineProps({
  currentPage: {
    required: true,
    type: Number,
  },
  defaultItemsPerPage: {
    default: undefined,
    type: Number,
  },
  totalItems: {
    required: true,
    type: Number,
  },
  inputDebounceTime: {
    default: undefined,
    type: Number,
  },
  itemsPerPageControlLabel: {
    default: "Items per page",
    type: String,
  },
  itemsPerPageOptions: {
    default: () => [5, 10, 20, 30],
    type: Array as () => number[],
  },
  pageInputLabel: {
    default: "Go to page:",
    type: String,
  },
  showFirstLastButtons: {
    default: true,
    type: Boolean,
  },
  showItemsPerPageControl: {
    default: true,
    type: Boolean,
  },
  showPageButtons: {
    default: false,
    type: Boolean,
  },
  showPageInput: {
    type: Boolean,
    default: false,
  },
  showPreviousNextButtons: {
    default: true,
    type: Boolean,
  },
});

const emit = defineEmits(["update:itemsPerPage", "update:currentPage"]);

const itemsPerPage = ref(
  props.defaultItemsPerPage || props.itemsPerPageOptions[0],
);

const lastPage = computed(() =>
  Math.ceil(props.totalItems / itemsPerPage.value),
);

const pages = computed(() =>
  Array.from({ length: lastPage.value }, (_, index) => index + 1),
);

const handleItemsPerPageChange = (event: Event) => {
  const newItemsPerPage = parseInt(
    (event.target as HTMLSelectElement).value,
    10,
  );
  itemsPerPage.value = newItemsPerPage;

  emit("update:itemsPerPage", itemsPerPage.value);
};

const handlePageInputChange = (value: string | number) => {
  const newPage = parseInt(value.toString(), 10) - 1;
  if (!isNaN(newPage) && newPage >= 0 && newPage < lastPage.value) {
    emit("update:currentPage", newPage);
  }
};
</script>

<style lang="css">
@import "../assets/css/pagination.css";
</style>
