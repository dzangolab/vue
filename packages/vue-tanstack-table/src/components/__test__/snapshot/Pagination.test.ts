import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import Pagination from "../../Pagination.vue";

describe("Pagination", () => {
  it("matches snapshot", () => {
    const wrapper = mount(Pagination, {
      global: {
        stubs: {
          DebounceInput: true,
        },
      },
      props: {
        currentPage: 1,
        showPageInput: true,
        showPageButtons: true,
        totalItems: 20,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
