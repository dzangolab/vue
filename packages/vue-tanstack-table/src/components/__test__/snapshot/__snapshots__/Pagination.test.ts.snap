// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > matches snapshot 1`] = `
<div
  class="pagination"
>
  <div
    class="items-per-page-control"
  >
    <span>
      Items per page
    </span>
    <select
      value="5"
    >
      
      <option
        value="5"
      >
        5
      </option>
      <option
        value="10"
      >
        10
      </option>
      <option
        value="20"
      >
        20
      </option>
      <option
        value="30"
      >
        30
      </option>
      
    </select>
  </div>
  <div
    class="page-input-control"
  >
    <span>
      Go to page:
    </span>
    <div
      class="field debounce-input"
    >
      <input
        aria-label="input"
        class="input-field"
        placeholder="Search..."
        type="number"
        value=""
      />
    </div>
  </div>
  <div
    class="pagination-buttons-wrapper"
  >
    <div>
      <button
        aria-label="button"
        class="button primary small filled icon-only first-page"
        role="button"
      >
        <span
          class="icon-left"
        >
          
          <svg
            aria-hidden="true"
            height="1em"
            role="img"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
          
        </span>
        <!---->
        <!---->
        <!---->
      </button>
      <button
        aria-label="button"
        class="button primary small filled icon-only previous-page"
        role="button"
      >
        <span
          class="icon-left"
        >
          
          <svg
            aria-hidden="true"
            height="1em"
            role="img"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
          
        </span>
        <!---->
        <!---->
        <!---->
      </button>
    </div>
    <div>
      
      <button
        aria-label="button"
        class="button primary small filled page-button"
        role="button"
      >
        <!---->
        <div
          class="label"
        >
          
          <span>
            1
          </span>
          
        </div>
        <!---->
        <!---->
      </button>
      <button
        aria-label="button"
        class="button primary small filled page-button active"
        role="button"
      >
        <!---->
        <div
          class="label"
        >
          
          <span>
            2
          </span>
          
        </div>
        <!---->
        <!---->
      </button>
      <button
        aria-label="button"
        class="button primary small filled page-button"
        role="button"
      >
        <!---->
        <div
          class="label"
        >
          
          <span>
            3
          </span>
          
        </div>
        <!---->
        <!---->
      </button>
      <button
        aria-label="button"
        class="button primary small filled page-button"
        role="button"
      >
        <!---->
        <div
          class="label"
        >
          
          <span>
            4
          </span>
          
        </div>
        <!---->
        <!---->
      </button>
      
    </div>
    <div>
      <button
        aria-label="button"
        class="button primary small filled icon-only next-page"
        role="button"
      >
        <span
          class="icon-left"
        >
          
          <svg
            aria-hidden="true"
            height="1em"
            role="img"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
          
        </span>
        <!---->
        <!---->
        <!---->
      </button>
      <button
        aria-label="button"
        class="button primary small filled icon-only last-page"
        role="button"
      >
        <span
          class="icon-left"
        >
          
          <svg
            aria-hidden="true"
            height="1em"
            role="img"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
          
        </span>
        <!---->
        <!---->
        <!---->
      </button>
    </div>
  </div>
</div>
`;
